include/lbann/callbacks/sync_selected.hpp

```diff
@@ -116,6 +143,29 @@ class sync_selected : public sync_layers {
   /// Synchronize and disble cuda profiler
   void do_sync(Layer* l) override;
 
+ private:
+
+  /** @class cuda_profiler_parameters
+   *  @brief Storage for parameters for the cuda profiler
+   */
+  struct cuda_profiler_parameters
+  {
+    std::string m_cfg_file;
+    std::string m_out_dir;
+    int m_out_mode;
+
+    cuda_profiler_parameters(const std::string& cfg_file,
+                             const std::string& out_dir,
+                             int out_mode)
+      : m_cfg_file(cfg_file),
+        m_out_dir(out_dir),
+        m_out_mode(out_mode)
+    {}
+  };// cuda_profiler_parameters
+
+  /** @brief The parameters for the CUDA profiler */
+  std::unique_ptr<cuda_profiler_parameters> m_cuda_prof_params;
```

Comment:
Why do we store this in a `unique_ptr`? I see later that we check whether it is `nullptr`, in which case we treat it as uninitialized and skip some stuff. Couldn't we do the same thing by setting a `bool`?

Reply:
1. Pointers are convertible to `bool`.
2. By using heap allocation, you don't unnecessarily allocate space for two unused strings and an `int`, in addition to the likely 4-8 extra bytes for the `bool`. Since the data is optionally provided, it makes sense to just throw it on the heap somewhere.